pisa-proxy/proxy/strategy/src/discovery/discovery.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 SphereEx Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    config::MasterHighAvailability,
+    monitors::{
+        connect_monitor::MonitorConnect, ping_monitor::MonitorPing,
+        read_only_monitor::MonitorReadOnly, replication_lag_monitor::MonitorReplicationLag,
+    },
+    readwritesplitting::ReadWriteEndpoint,
+};
+
+//define discovery kind (support MHA,RDS,MGR etc.)
+pub enum DiscoveryKind {
+    MasterHighAvailability(DiscoveryMasterHighAvailability),
+}
+
+pub trait Discovery {
+    type Output;
+
+    fn build(config: MasterHighAvailability, rw_endpoint: ReadWriteEndpoint) -> Self::Output;
+    fn build_monitors(
+        &self,
+        monitor_response_channel: crate::readwritesplitting::MonitorResponseChannel,
+    ) -> Vec<MonitorKind>;
+}
+
+pub struct DiscoveryMasterHighAvailability {
+    config: MasterHighAvailability,
+    rw_endpoint: ReadWriteEndpoint,
+    pub monitors: Vec<MonitorKind>,
+}
+
+impl Discovery for DiscoveryMasterHighAvailability {
+    type Output = Self;
+
+    fn build(config: MasterHighAvailability, rw_endpoint: ReadWriteEndpoint) -> Self::Output {
+        Self { config, rw_endpoint, monitors: vec![] }
+    }
+
+    fn build_monitors(
+        &self,
+        monitor_response_channel: crate::readwritesplitting::MonitorResponseChannel,
+    ) -> Vec<MonitorKind> {
+        let mut monitors = vec![];
```

Comment:
Could make this a factory in next release.

Reply:
mark